From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>
Date: Thu, 2 Mar 2023 17:33:29 +0900
Subject: bpf: Ignore HOST_ID resolved from ipcache for IPv6 case

[ upstream commit a74affd5c4851de3adface98a0b81bcca9ffde55 ]

This PR adds the check to ignore HOST_ID resolved from ipcache for IPv6 case.
resolve_srcid_ipv6 should ignore the HOST_ID resolved from ipcache because the
packets marked MARK_MAGIC_HOST actually be from the host.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>
Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
---
 bpf/bpf_host.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/bpf/bpf_host.c b/bpf/bpf_host.c
index a43a8847d3..806ee4e2ae 100644
--- a/bpf/bpf_host.c
+++ b/bpf/bpf_host.c
@@ -160,8 +160,20 @@ resolve_srcid_ipv6(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
 	if (identity_is_reserved(srcid_from_ipcache)) {
 		src = (union v6addr *) &ip6->saddr;
 		info = lookup_ip6_remote_endpoint(src);
-		if (info != NULL && info->sec_label)
-			srcid_from_ipcache = info->sec_label;
+		if (info) {
+			if (info->sec_label) {
+				/* When SNAT is enabled on traffic ingressing
+				 * into Cilium, all traffic from the world will
+				 * have a source IP of the host. It will only
+				 * actually be from the host if "srcid_from_proxy"
+				 * (passed into this function) reports the src as
+				 * the host. So we can ignore the ipcache if it
+				 * reports the source as HOST_ID.
+				 */
+				if (info->sec_label != HOST_ID)
+					srcid_from_ipcache = info->sec_label;
+			}
+		}
 		cilium_dbg(ctx, info ? DBG_IP_ID_MAP_SUCCEED6 : DBG_IP_ID_MAP_FAILED6,
 			   ((__u32 *) src)[3], srcid_from_ipcache);
 	}
-- 
2.39.5 (Apple Git-154)

